### PR TITLE
conducting interview

### PR DESCRIPTION
**Pull request by vvs79**
code review:

- [ ] @DimaSmagulov 
- [ ] @krisberry 
- [ ] @natalya5 
- [ ] @oleg08 
- [ ] @OrestF 
- [ ] @rtershak 
- [ ] @CHILL99 

2nd lvl review:

- [ ] @Kmalynss
- [ ] @serhiym 
- [ ] @ruba-ruba

### CODE REVIEW

**Comment by vvs79** [comment]
https://trello.com/c/2XrWHrc4/51-as-an-interviewer-i-want-to-be-able-to-go-through-a-list-of-questions-when-conducting-interview

**Comment by vvs79** [comment]
q

File: app/controllers/interview_questions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+include ActionView::Helpers::TextHelper
+
+class InterviewQuestionsController < ApplicationController
+  before_action :authenticate_user!
+  before_action :check_interviewer
+  before_action :check_template
+  add_breadcrumb "interviews", :interviews_path
+
+  def editmultiple
```
Comment (ruba-ruba) [inline app/controllers/interview_questions_controller.rb]: please add  `underscore`  between words

File: app/controllers/interview_questions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    @interview = Interview.last
+    add_breadcrumb "workflow_" + (@interview.id).to_s, editmultiple_interview_questions_path
+    @interview.template.questions.each do |q|
+      InterviewQuestion.create(interview_id: @interview.id, question_id: q.id)
+    end
+  end
+
+  def updatemultiple
```
Comment (ruba-ruba) [inline app/controllers/interview_questions_controller.rb]: same as above

File: app/controllers/interview_questions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    @interview = Interview.last
+    @interviewquestions = InterviewQuestion.update(params[:interviewquestions].keys, params[:interviewquestions].values).reject { |p| p.errors.empty? }
+    redirect_to interview_path(@interview)
+  end
+
+  private
+
+  def interviewquestion_params
```
Comment (ruba-ruba) [inline app/controllers/interview_questions_controller.rb]: same as above

File: app/controllers/interview_questions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+class InterviewQuestionsController < ApplicationController
+  before_action :authenticate_user!
+  before_action :check_interviewer
+  before_action :check_template
+  add_breadcrumb "interviews", :interviews_path
+
+  def editmultiple
+    @interview = Interview.last
+    add_breadcrumb "workflow_" + (@interview.id).to_s, editmultiple_interview_questions_path
+    @interview.template.questions.each do |q|
```
Comment (ruba-ruba) [inline app/controllers/interview_questions_controller.rb]: i guess you can write `@interview.questions`

File: app/controllers/interview_questions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      InterviewQuestion.create(interview_id: @interview.id, question_id: q.id)
+    end
+  end
+
+  def updatemultiple
+    @interview = Interview.last
```
Comment (ruba-ruba) [inline app/controllers/interview_questions_controller.rb]: why last ?

File: app/controllers/interview_questions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    add_breadcrumb "workflow_" + (@interview.id).to_s, edit_multiple_interview_questions_path
+    @interview.questions.each do |q|
+      InterviewQuestion.create(interview_id: @interview.id, question_id: q.id)
+    end
+  end
+
+  def update_multiple
+    @interview = Interview.last
+    @interviewquestions = InterviewQuestion.update(params[:interview_questions].keys, params[:interview_questions].values).reject { |p| p.errors.empty? }
```
Comment (ruba-ruba) [inline app/controllers/interview_questions_controller.rb]: :x:

File: app/views/interview_questions/edit_multiple.html.slim [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  h1.center
+	  | Interview: 
+	  = @interview.id
+  br
+  .form
+    = form_tag update_multiple_interview_questions_path, method: :put do 
+      #tabs
+        ul
+          - num=0
+          - @interview.questions.each do |q|
```
Comment (ruba-ruba) [inline app/views/interview_questions/edit_multiple.html.slim]: if you need index use `each_with_index`

File: app/controllers/interview_questions_controller.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
+include ActionView::Helpers::TextHelper
+
+class InterviewQuestionsController < ApplicationController
+  before_action :authenticate_user!
+  before_action :check_interviewer
+  before_action :check_template
+  add_breadcrumb "interviews", :interviews_path
+
+  def edit_multiple
+    @interview = Interview.last
```
Comment (ruba-ruba) [inline app/controllers/interview_questions_controller.rb]: `Interview.last` is not correct

File: app/views/interview_questions/edit_multiple.html.slim (hunk 7 above)
Comment (vvs79) [inline app/views/interview_questions/edit_multiple.html.slim]: need only one last index

File: app/controllers/interview_questions_controller.rb (hunk 8 above)
Comment (vvs79) [inline app/controllers/interview_questions_controller.rb]: need only one last index

File: app/views/interview_questions/edit_multiple.html.slim [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+            - num+=1
+            - l="li"+num.to_s
+            - hrf="#tabs-"+num.to_s
+            li.d-line id=l
+              a href=hrf
+                = num 
+        br
+        - num_tab=0
+        - @interview.questions.each do |q|
```
Comment (Kmalynss) [inline app/views/interview_questions/edit_multiple.html.slim]: ```each_with_index```

File: app/views/interview_questions/edit_multiple.html.slim [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+            h3
+              | Question 
+              = num_tab+1
+              | :  
+              = q.content
+            h4
+              | Answer: 
+              = q.answer
+            br
+            - interview_question = InterviewQuestion.where(interview_id: @interview.id, question_id: q.id).first
```
Comment (Kmalynss) [inline app/views/interview_questions/edit_multiple.html.slim]: use relations and instance methods for this.
P.S. this is not first time when i remind you to use relations

File: app/controllers/interview_questions_controller.rb (hunk 6 above)
Comment (vvs79) [inline app/controllers/interview_questions_controller.rb]: did it early

File: app/views/interview_questions/edit_multiple.html.slim (hunk 10 above)
Comment (Kmalynss) [inline app/views/interview_questions/edit_multiple.html.slim]: BTW any kind of assigning and defining of any kind of variables in views is not acceptable

File: app/views/interviews/show.html.slim [hunk 11]
```diff
@@ -38,8 +38,9 @@
             |#{q.id}
   - @questions.each do |q|
     .summary-card id="question#{q.id}"
+      - iq = InterviewQuestion.where(interview_id: @interview.id, question_id: q.id).first
```
Comment (Kmalynss) [inline app/views/interviews/show.html.slim]: use relations and instance methods for this.
P.S. this is not first time when i remind you to use relations

File: app/assets/javascripts/workflow.js.erb [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+  $("input.color_radio").click(function() {
+    var color_radio = $(this).attr("value");
+    if (color_radio=='green') { color='green'; }
+    if (color_radio=='yellow') { color='yellow'; }
+    if (color_radio=='red') { color='red'; }
+    $("a#ui-id-"+num_i).css("background",color);
+  });
+  
+});
```
Comment (Kmalynss) [inline app/assets/javascripts/workflow.js.erb]: do we really need this js?

File: app/controllers/interview_questions_controller.rb [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+  private
+
+  def interview_question_params
+    params.require(:interview_question).permit(:interview_id, :question_id, :comment, :mark)
+  end
+
+  def check_interviewer
+    return true unless current_user.admin?
+    redirect_to authenticated_root_path, notice: 'Access Denied'
```
Comment (Kmalynss) [inline app/controllers/interview_questions_controller.rb]: why not? 
```ruby
 redirect_to authenticated_root_path, notice: 'Access Denied' if current_user.admin?
```

File: app/controllers/interview_questions_controller.rb (hunk 6 above)
Comment (vvs79) [inline app/controllers/interview_questions_controller.rb]: corrected it

File: app/controllers/interview_questions_controller.rb (hunk 8 above)
Comment (Kmalynss) [inline app/controllers/interview_questions_controller.rb]: >>> need only one last index
This doesn't make any sense.
BTW what will happen if 2 interviews will conducted in a same moment?

**Comment by Kmalynss** [comment]
Why there is no tests?

**Comment by Kmalynss** [comment]
Also, creating one big form for entire interview isn't good approach.
Each question should be represented with its own page. User after submit should be redirected to a next question.

File: app/assets/javascripts/workflow.js.erb (hunk 12 above)
Comment (vvs79) [inline app/assets/javascripts/workflow.js.erb]: with this code better looking

File: app/controllers/interview_questions_controller.rb (hunk 13 above)
Comment (vvs79) [inline app/controllers/interview_questions_controller.rb]: thanks for the tip

File: app/views/interview_questions/edit_multiple.html.slim [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+            h3
+              | Question 
+              = num_tab+1
+              | :  
+              = q.content
+            h4
+              | Answer: 
+              = q.answer
+            br
+            = fields_for "interview_questions[]", InterviewQuestion.where(interview_id: @interview.id, question_id: q.id).first do |f|
```
Comment (Kmalynss) [inline app/views/interview_questions/edit_multiple.html.slim]: @vvs79 if you don't understand my comment, please ask me via skype.
why didn't you use:
```ruby
@interview.interview_questions.each do |interview_question|
  ...
  = fields_for "interview_questions[]", interview_question
```
Your current change does not correlates with my comment and looks like an attempt to trick me =)

**Comment by vvs79** [comment]
https://trello.com/b/gEb5Fzwm/aid-interview

**Comment by Kmalynss** [comment]
As we where talking after the demo. Each question should be on its own page. After submit user should be redirect to next question.

File: app/controllers/interview_questions_controller.rb [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    @interview = Interview.where(user_id: current_user.id).last
+    @interview.interview_questions.each do |f|
+      if f.mark.blank?
+        f.update(interview_question_params)
+        break
+      end
+    end
+    redirect_to :back
+  end
+
```
Comment (Kmalynss) [inline app/controllers/interview_questions_controller.rb]: this code isn't good

there should not be actions like ```edit_multiple``` and ```update_multiple```

File: app/views/interview_questions/edit.html.slim [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    | Please select mark
+  br
+  - a = 0
+  - b = 1
+  - @interview.interview_questions.each_with_index do |q, i|
+    - a += 1
+    - b = i + 1 if q.question_id == @interview_question.question_id
+  - @questions = @interview.questions
+  - @question = @questions.find(@interview_question.question_id)
+  h3
```
Comment (Kmalynss) [inline app/views/interview_questions/edit.html.slim]: any kind of variables are not allowed to be defined in a view

File: app/views/interview_questions/edit.html.slim [hunk 17]
```diff
@@ -0,0 +1,46 @@
+- provide(:title, 'Interview questions')
+.container.content-area
+  h1.center
+    | Interview: 
+    = @interview.id
+  p.text-info
+    | Please select mark
+  br
+  - a = 0
+  - b = 1
```
Comment (Kmalynss) [inline app/views/interview_questions/edit.html.slim]: variable name must be expressive
```a, b``` can not be used

File: app/views/interview_questions/index.html.slim [hunk 18]
```diff
@@ -0,0 +1,24 @@
+- provide(:title, 'All interview questions')
+.container.content-area
+  h1
+    | All interview questions
+  .row  
+    = render 'layouts/messages'
+  br
+  table.table.table-bordered.table-hover
+    tr.active
+      th width="5%"
```
Comment (Kmalynss) [inline app/views/interview_questions/index.html.slim]: tag attributes can not be used for changing layout.
Use css instead. Find and fix such mistakes in this PR.

File: app/views/interviews/_interview.html.slim [hunk 19]
```diff
@@ -25,4 +25,14 @@
...
       = link_to interview_path(i), method: :delete, data: { confirm: 'Are you sure?' }, class: "btn btn-sm btn-danger" do
         i.glyphicon.glyphicon-remove
-
+    td
+      = link_to edit_interview_interview_question_path(i, i.interview_questions.first), class: "btn btn-sm btn-info"
+        - m = 0
+        - a = 0
+        -i.interview_questions.each do |q|
+          - m +=1 if (q.mark).nil?
+          - a += 1
```
Comment (Kmalynss) [inline app/views/interviews/_interview.html.slim]: Any kind of logic must be avoided in a view. 
This must be refactored. Consider helpers, controllers and models as possible place for such code.

File: app/controllers/interviews_controller.rb [hunk 20]
```diff
@@ -22,17 +25,25 @@ def show
 
   def create
     @interview = Interview.new(interview_params)
-    @questions = @interview.questions
     if @interview.save
-      redirect_to interview_path(@interview), notice: 'Interview was successfully created'
+      @interview.questions.each do |q|
+        InterviewQuestion.create(interview_id: @interview.id, question_id: q.id)
```
Comment (Kmalynss) [inline app/controllers/interviews_controller.rb]: again, use relations.
```ruby
@interview,interview_questions.create
```

**Comment by Kmalynss** [comment]
please rebase, to get new travis config

**Pull request closed without merging**